File: samples/bot-configuration-app-auth/csharp/Bot configuration/Bots/TeamsBot.cs
using AdaptiveCards;
using Bogus.DataSets;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Teams;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Schema.Teams;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Net.Http;
using System.Net;
using System.Threading;
using System.Threading.Tasks;
using System.IO;

namespace Botconfiguration.Bots
{
    /// <summary>
    /// TeamsBot class handles bot activities for Microsoft Teams.
    /// </summary>
    public class TeamsBot : TeamsActivityHandler
    {
        private string _chosenFlow = string.Empty;

        /// <summary>
        /// Handles the event when members are added to the channel.
        /// </summary>
        /// <param name="membersAdded">List of members added.</param>
        /// <param name="turnContext">Turn context.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        protected override async Task OnMembersAddedAsync(IList<ChannelAccount> membersAdded, ITurnContext<IConversationUpdateActivity> turnContext, CancellationToken cancellationToken)
        {
            const string imagePath = "Images/configbutton.png";
            var imageData = Convert.ToBase64String(await File.ReadAllBytesAsync(imagePath));

            var adaptiveCardJson = $@"
                {{
                    ""$schema"": ""http://adaptivecards.io/schemas/adaptive-card.json"",
                    ""type"": ""AdaptiveCard"",
                    ""version"": ""1.0"",
                    ""body"": [
                        {{
                            ""type"": ""TextBlock"",
                            ""text"": ""Hello and welcome! With this sample, you can experience the functionality of bot configuration. To access Bot configuration, click on the settings button in the bot description card."",
                            ""wrap"": true,
                            ""size"": ""large"",
                            ""weight"": ""bolder""
                        }},
                        {{
                            ""type"": ""Image"",
                            ""url"": ""data:image/png;base64,{imageData}"",
                            ""size"": ""auto""
                        }}
                    ],
                    ""fallbackText"": ""This card requires Adaptive Card support.""
                }}";

            var attachment = new Attachment
            {
                ContentType = AdaptiveCard.ContentType,
                Content = JsonConvert.DeserializeObject(adaptiveCardJson)
            };

            var reply = MessageFactory.Attachment(attachment);
            await turnContext.SendActivityAsync(reply, cancellationToken);
        }

        /// <summary>
        /// Handles the event when a message activity is received.
        /// </summary>
        /// <param name="turnContext">Turn context.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
            var activity = turnContext.Activity;
            if (!string.IsNullOrEmpty(activity.Text))
            {
                var text = activity.Text.ToLower().Trim();
                if (text == "chosen flow" || text == "<at>typeahead search adaptive card</at> chosen flow")
                {
                    await turnContext.SendActivityAsync($"Bot configured for {_chosenFlow} flow", cancellationToken: cancellationToken);
                }
            }
            else if (activity.Value != null)
            {
                await turnContext.SendActivityAsync($"Selected option is: {activity.Value}", cancellationToken: cancellationToken);
            }
        }

        /// <summary>
        /// Handles the event when a configuration fetch is requested.
        /// </summary>
        /// <param name="turnContext">Turn context.</param>
        /// <param name="configData">Configuration data.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>Configuration response.</returns>
        protected override Task<ConfigResponseBase> OnTeamsConfigFetchAsync(ITurnContext<IInvokeActivity> turnContext, JObject configData, CancellationToken cancellationToken)
        {
            var response = new ConfigResponse<BotConfigAuth>
            {
                Config = new BotConfigAuth
                {
                    SuggestedActions = new SuggestedActions
                    {
                        Actions = new List<CardAction>
                            {
                                new CardAction
                                {
                                    Type = ActionTypes.OpenUrl,
                                    Title = "Sign in to this app",
                                    Value = "https://example.com/auth"
                                }
                            }
                    },
                    Type = "auth"
                }
            };
            return Task.FromResult<ConfigResponseBase>(response);
        }

        /// <summary>
        /// Handles the event when a configuration submit is requested.
        /// </summary>
        /// <param name="turnContext">Turn context.</param>
        /// <param name="configData">Configuration data.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>Configuration response.</returns>
        protected override Task<ConfigResponseBase> OnTeamsConfigSubmitAsync(ITurnContext<IInvokeActivity> turnContext, JObject configData, CancellationToken cancellationToken)
        {
            var response = new ConfigResponse<TaskModuleResponseBase>
            {
                Config = new TaskModuleMessageResponse
                {
                    Type = "message",
                    Value = "You have chosen to finish setting up bot"
                }
            };

            return Task.FromResult<ConfigResponseBase>(response);
        }
    }
}

File: samples/TeamsSDK/Archived/bot-configuration-app-auth/nodejs/teamsBot.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { TeamsActivityHandler, CardFactory } = require("botbuilder");
const fs = require('fs');

/**
 * TeamsBot class extends TeamsActivityHandler to handle bot activities.
 */
class TeamsBot extends TeamsActivityHandler {
  constructor() {
    super();
    this.onMembersAdded(this.handleMembersAdded.bind(this));
    this.onMessage(this.handleMessage.bind(this));
  }

  /**
   * Handles the event when new members are added to the conversation.
   * @param {TurnContext} context - The context object for the turn.
   * @param {Function} next - The next middleware handler to call.
   */
  async handleMembersAdded(context, next) {
    const imagePath = 'Images/configbutton.png';
    const imageBase64 = fs.readFileSync(imagePath, 'base64');
    const card = CardFactory.adaptiveCard({
      "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
      "type": "AdaptiveCard",
      "version": "1.0",
      "body": [
        {
          "type": "TextBlock",
          "text": "Hello and welcome! With this sample, you can experience the functionality of bot configuration. To access Bot configuration, click on the settings button in the bot description card.",
          "wrap": true,
          "size": "large",
          "weight": "bolder"
        },
        {
          "type": "Image",
          "url": `data:image/jpeg;base64,${imageBase64}`,
          "size": "auto"
        }
      ],
      "fallbackText": "This card requires Adaptive Card support."
    });

    await context.sendActivity({
      text: '',
      attachments: [card]
    });

    await next();
  }

  /**
   * Handles incoming messages.
   * @param {TurnContext} context - The context object for the turn.
   * @param {Function} next - The next middleware handler to call.
   */
  async handleMessage(context, next) {
    // By calling next() you ensure that the next BotHandler is run.
    await next();
  }

  /**
   * Handles the bot configuration fetch request.
   * @param {TurnContext} _context - The context object for the turn.
   * @param {Object} _configData - The configuration data.
   * @returns {Object} The response object containing the configuration.
   */
  async handleTeamsConfigFetch(_context, _configData) {
    return {
      config: {
        type: "auth",
        suggestedActions: {
          actions: [
            {
              type: "openUrl",
              value: "https://example.com/auth",
              title: "Sign in to this app"
            }
          ]
        }
      }
    };
  }

  /**
   * Handles the bot configuration submit request.
   * @param {TurnContext} context - The context object for the turn.
   * @param {Object} _configData - The configuration data.
   * @returns {Object} The response object containing the configuration.
   */
  async handleTeamsConfigSubmit(context, _configData) {
    return {
      config: {
        type: 'message',
        value: 'You have chosen to finish setting up bot',
      }
    };
  }
}

module.exports.TeamsBot = TeamsBot;